File: sdk/go/validation.go
package omnidex

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"net/url"
	"path"
	"strings"
	"unicode/utf8"
)

const (
	maxPromptBytes   = 4 * 1024
	maxResponseBytes = 16 * 1024 * 1024
)

func validateClientConfiguration(baseURL, token string) error {
	if baseURL == "" || baseURL != strings.TrimSpace(baseURL) {
		return fmt.Errorf("Omnidex base URL must be exact nonblank text")
	}
	parsed, err := url.Parse(baseURL)
	if err != nil || (parsed.Scheme != "http" && parsed.Scheme != "https") || parsed.Host == "" ||
		parsed.User != nil || parsed.RawQuery != "" || parsed.Fragment != "" {
		return fmt.Errorf("Omnidex base URL must be one HTTP(S) origin or canonical base path")
	}
	if parsed.Path != "" && (path.Clean(parsed.Path) != parsed.Path || strings.HasSuffix(parsed.Path, "/")) {
		return fmt.Errorf("Omnidex base URL path must be canonical without a trailing slash")
	}
	if err := validateToken(token); err != nil {
		return err
	}
	return nil
}

func validateToken(token string) error {
	if len(token) < 32 || len(token) > 4096 {
		return fmt.Errorf("Omnidex integration token must contain 32..4096 exact ASCII bytes")
	}
	for _, character := range []byte(token) {
		if character < 0x21 || character > 0x7e {
			return fmt.Errorf("Omnidex integration token must contain only visible ASCII bytes")
		}
	}
	return nil
}

func validateCanonicalID(label, value string, maximum int) error {
	if value == "" || len(value) > maximum {
		return fmt.Errorf("%s must contain 1..%d canonical bytes", label, maximum)
	}
	for index, character := range []byte(value) {
		allowed := character >= 'a' && character <= 'z' || character >= '0' && character <= '9' ||
			character == '_' || character == '.' || character == ':' || character == '-'
		if !allowed || index == 0 && !((character >= 'a' && character <= 'z') ||
			(character >= '0' && character <= '9')) {
			return fmt.Errorf("%s is not canonical", label)
		}
	}
	return nil
}

func ValidateDelegatedAuthorityID(value string) error {
	if len(value) != 68 || !strings.HasPrefix(value, "dba_") {
		return fmt.Errorf("delegated authority must be one opaque dba_ identity")
	}
	decoded, err := hex.DecodeString(value[4:])
	if err != nil || len(decoded) != 32 || value[4:] != strings.ToLower(value[4:]) {
		return fmt.Errorf("delegated authority must be one opaque dba_ identity")
	}
	return nil
}

func NewDelegatedAuthorityID() (string, error) {
	value := make([]byte, 32)
	if _, err := rand.Read(value); err != nil {
		return "", fmt.Errorf("generate delegated authority identity: %w", err)
	}
	return "dba_" + hex.EncodeToString(value), nil
}

func validatePrompt(value string) error {
	if strings.TrimSpace(value) == "" || len(value) > maxPromptBytes || !utf8.ValidString(value) ||
		strings.ContainsRune(value, '\x00') {
		return fmt.Errorf("prompt must contain 1..%d exact nonblank UTF-8 bytes without NUL", maxPromptBytes)
	}
	return nil
}

File: sdk/javascript/src/validation.js
const CANONICAL_ID = /^[a-z0-9][a-z0-9_.:-]*$/;
const AUTHORITY_ID = /^dba_[0-9a-f]{64}$/;
const ENVIRONMENT_NAME = /^OMNIDEX_DELEGATED_AUTHORITY_[A-Z][A-Z0-9_]{0,93}_TOKEN$/;
const SSL_MODES = new Set(["disable", "allow", "prefer", "require", "verify-ca", "verify-full"]);

export function validateConfiguration(baseUrl, token) {
  exactString(baseUrl, "Omnidex base URL");
  let parsed;
  try {
    parsed = new URL(baseUrl);
  } catch {
    throw new TypeError("Omnidex base URL must be one HTTP(S) origin or canonical base path.");
  }
  if (!["http:", "https:"].includes(parsed.protocol) || parsed.username || parsed.password || parsed.search || parsed.hash || baseUrl.endsWith("/")) {
    throw new TypeError("Omnidex base URL must be one HTTP(S) origin or canonical base path without a trailing slash.");
  }
  validateToken(token);
}

export function validateToken(token) {
  if (typeof token !== "string" || token.length < 32 || token.length > 4096 || !/^[\x21-\x7e]+$/.test(token)) {
    throw new TypeError("Omnidex integration token must contain 32..4096 exact visible ASCII bytes.");
  }
}

export function validateCanonicalId(label, value, maximum) {
  if (typeof value !== "string" || value.length < 1 || value.length > maximum || !CANONICAL_ID.test(value)) {
    throw new TypeError(`${label} is not canonical.`);
  }
}

export function validateAuthorityId(value) {
  if (typeof value !== "string" || !AUTHORITY_ID.test(value)) {
    throw new TypeError("Delegated authority must be one opaque dba_ identity.");
  }
}

export function validatePrompt(value) {
  if (typeof value !== "string" || !value.trim() || new TextEncoder().encode(value).length > 4096 || value.includes("\0")) {
    throw new TypeError("Prompt must contain 1..4096 exact nonblank UTF-8 bytes without NUL.");
  }
}

export function validateDirectDataSource(input) {
  exactString(input?.name, "Data-source name");
  if (!Number.isSafeInteger(input.port) || input.port < 1 || input.port > 65535) {
    throw new TypeError("PostgreSQL port must be between 1 and 65535.");
  }
  if (!SSL_MODES.has(input.sslMode)) throw new TypeError(`PostgreSQL SSL mode ${JSON.stringify(input.sslMode)} is unsupported.`);
  if (input.useDsn) {
    exactString(input.dsn, "PostgreSQL DSN");
  } else {
    exactString(input.host, "PostgreSQL host");
    exactString(input.databaseName, "PostgreSQL database");
    exactString(input.username, "PostgreSQL username");
  }
}

export function validateDelegatedDataSource(input) {
  exactString(input?.name, "Data-source name");
  validateConfiguration(input.authorityUrl, "12345678901234567890123456789012");
  if (!ENVIRONMENT_NAME.test(input.credentialEnv ?? "")) {
    throw new TypeError("Credential environment variable is outside the dedicated namespace OMNIDEX_DELEGATED_AUTHORITY_*.");
  }
}

export function validateChannel(input) {
  validateCanonicalId("Channel ID", input?.id, 96);
  validateCanonicalId("Data-source ID", input?.dataSourceId, 128);
  exactString(input.name, "Channel name");
  exactString(input.workspaceRoot, "Channel workspace root");
  if (!Array.isArray(input.tags) || input.tags.length > 32) throw new TypeError("Channel tags must be an array with at most 32 entries.");
  const seen = new Set();
  for (const tag of input.tags) {
    if (typeof tag !== "string" || !tag || tag !== tag.trim() || tag !== tag.toLowerCase() || tag.length > 64 || seen.has(tag)) {
      throw new TypeError(`Channel tag ${JSON.stringify(tag)} is not canonical and unique.`);
    }
    seen.add(tag);
  }
}

function exactString(value, label) {
  if (typeof value !== "string" || !value || value !== value.trim() || value.includes("\0")) {
    throw new TypeError(`${label} must be exact nonblank text.`);
  }
}
